File: src/main/java/org/keynote/godtools/android/content/LanguagesLoader.java
package org.keynote.godtools.android.content;

import android.content.Context;
import android.support.annotation.NonNull;

import org.ccci.gto.android.common.db.Query;
import org.ccci.gto.android.common.eventbus.content.CachingAsyncTaskEventBusLoader;
import org.cru.godtools.model.Language;
import org.cru.godtools.model.event.content.LanguageEventBusSubscriber;
import org.keynote.godtools.android.db.GodToolsDao;

import java.util.List;

public final class LanguagesLoader extends CachingAsyncTaskEventBusLoader<List<Language>> {
    private static final Query<Language> QUERY_LANGUAGES = Query.select(Language.class);

    @NonNull
    private final GodToolsDao mDao;

    public LanguagesLoader(@NonNull final Context context) {
        super(context);
        mDao = GodToolsDao.getInstance(context);

        addEventBusSubscriber(new LanguageEventBusSubscriber(this));
    }

    @Override
    public List<Language> loadInBackground() {
        return mDao.get(QUERY_LANGUAGES);
    }
}

File: app/src/main/java/org/cru/godtools/content/LanguagesLoader.kt
package org.cru.godtools.content

import android.content.Context
import org.ccci.gto.android.common.db.Query
import org.ccci.gto.android.common.eventbus.content.CachingAsyncTaskEventBusLoader
import org.cru.godtools.model.Language
import org.cru.godtools.model.event.content.LanguageEventBusSubscriber
import org.keynote.godtools.android.db.Contract.LanguageTable
import org.keynote.godtools.android.db.Contract.TranslationTable
import org.keynote.godtools.android.db.GodToolsDao

internal val QUERY_LANGUAGES = Query.select(Language::class.java)
    .join(LanguageTable.SQL_JOIN_TRANSLATION)
    .where(TranslationTable.SQL_WHERE_PUBLISHED)

class LanguagesLoader(context: Context) :
    CachingAsyncTaskEventBusLoader<List<@JvmSuppressWildcards Language>>(context) {
    private val dao = GodToolsDao.getInstance(context)

    init {
        addEventBusSubscriber(LanguageEventBusSubscriber(this))
    }

    override fun loadInBackground(): List<Language> = dao.get(QUERY_LANGUAGES)
}
